### PR TITLE
[RISCV] Support RISCV Atomics ABI attributes

### PR DESCRIPTION
**Pull request by ilovepi**
This patch adds support for the `atomic_abi` attribute, specifid in
https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-elf.adoc#tag_riscv_atomic_abi-14-uleb128version.

The atomics_abi tag merging is conducted as follows:

- UNKNOWN is safe to merge with all other values.
- A6C is compatible with A6S, and results in the A6C ABI.
- A6C is incompatible with A7, and results in an error.
- A6S and A7 are compatible, and merging results in the A7 ABI.

Note: the A7 is not yet supported in either LLVM or in any current hardware,
and is therefore ommited from attribute generation in RISCVTargetStreamer.

### CODE REVIEW

File: lld/ELF/Arch/RISCV.cpp
Comment (petrhosek) [inline lld/ELF/Arch/RISCV.cpp:1092]: Leftover debug print?

File: lld/ELF/Arch/RISCV.cpp
Comment (petrhosek) [inline lld/ELF/Arch/RISCV.cpp:1140]: Where is this function used?

File: lld/ELF/Arch/RISCV.cpp
Comment (ilovepi) [reply lld/ELF/Arch/RISCV.cpp:1092]: Oh I missed that. Thanks.

File: lld/ELF/Arch/RISCV.cpp
Comment (ilovepi) [reply lld/ELF/Arch/RISCV.cpp:1140]: Shoot this belongs in another PR. I’ll update these in the morning. Thanks for pointing that out.

File: lld/ELF/Arch/RISCV.cpp
Comment (ilovepi) [reply lld/ELF/Arch/RISCV.cpp]: done.

File: lld/ELF/Arch/RISCV.cpp
Comment (topperc) [inline lld/ELF/Arch/RISCV.cpp:1099]: Can we use the original iterator returned by try_emplace to avoid the second map lookup here?

Comment (ilovepi) [reply lld/ELF/Arch/RISCV.cpp]: done.

**Comment by ilovepi** [comment]
@jrtc27 Do you have thoughts on this change w.r.t. the defaults? I recall you had expressed concerns previously in the psABI, so I want to be sure we're changing the status quo appropriately.

**Comment by jrtc27** [comment]
> @jrtc27 Do you have thoughts on this change w.r.t. the defaults? I recall you had expressed concerns previously in the psABI, so I want to be sure we're changing the status quo appropriately.

To be clear, without reading the patch in detail, this is making the default be A6S? I don't think I have concerns with that other than to note that transitioning from A6S to A7 will be an ABI break for existing binaries, so the timeline for when that's ok to do is unclear.

**Comment by ilovepi** [comment]
ping. I think everything has been addressed.

File: llvm/lib/Target/RISCV/MCTargetDesc/RISCVTargetStreamer.cpp
Comment (asb) [inline llvm/lib/Target/RISCV/MCTargetDesc/RISCVTargetStreamer.cpp:81]: Is this the right way round? Isn't `STI.hasFeature(RISCV::FeatureTrailingSeqCstFence)` true if `+no-seq-cst-trailing-fence` is present? And if there's no trailing fence, then we have A6C? Or am I misreading?

I think it would be better to name the feature `FeatureNoTrailingSeqCstFence` as we do for `FeatureNoRVCHints`. Though I think this patch would be best if it aimed to change no defaults at all, and left that as a follow-on patch (which would hopefully be a quick LGTM as I think there's no objection to that, going by this review thread).

Comment (asb) [reply llvm/lib/Target/RISCV/MCTargetDesc/RISCVTargetStreamer.cpp:81]: > I think this should be correct. The default is now "enabled", so adding `+no-seq-cst-trailing-fence` should make `STI.hasFeature(RISCV::FeatureTrailingSeqCstFence)` return `false`. I think the tests indicate that's how its working.

So in attributes.ll, the A6S output has the RUN line `; RUN: llc -mtriple=riscv64 -mattr=+a,+no-seq-cst-trailing-fence %s -o - | FileCheck --check-prefixes=CHECK,RV64A,A6S %s`. But for A6S, don't we require that the trailing fence is emitted?

Comment (ilovepi) [reply llvm/lib/Target/RISCV/MCTargetDesc/RISCVTargetStreamer.cpp:81]: huh, you're right. Let me double check that now that the default is in it's own change.

**Comment by asb** [comment]
Thanks for splitting out that change. I'm thinking you might be missing a test case in lld/test/ELF/riscv-attributes.s. You handle an input file where the atomic_abi attribute is present and explicitly set to `UNKNOWN`. But what about when merging a file where the attribute was explicitly set, with one where the attribute wasn't present at all?

**Comment by ilovepi** [comment]
> Thanks for splitting out that change. I'm thinking you might be missing a test case in lld/test/ELF/riscv-attributes.s. You handle an input file where the atomic_abi attribute is present and explicitly set to `UNKNOWN`. But what about when merging a file where the attribute was explicitly set, with one where the attribute wasn't present at all?

Oh, that's a good catch. I didn't think about that case. Let me see what happens.

File: lld/test/ELF/riscv-attributes.s
Comment (asb) [inline lld/test/ELF/riscv-attributes.s:66]: Shouldn't the check-prefix for this be NONE_UNKNOWN to match the convention elsewhere?

File: lld/test/ELF/riscv-attributes.s
Comment (asb) [inline lld/test/ELF/riscv-attributes.s:68]: Shouldn't the output filename be atomic_abi_A6C_none?

File: lld/test/ELF/riscv-attributes.s
Comment (asb) [inline lld/test/ELF/riscv-attributes.s:335]: My other two comments were minor nits, this is the one where I'm wondering if there's some missing logic.

For every other case where one of the input files has the atomic_abi attribute and we produce an output, the output file also has the attribute. In this case, although one of the inputs had the attribute we end up dropping it in the output. I think that's probably not desired, but maybe I'm missing a reason to do this?

**Review by asb — COMMENTED** [review]
One question in line about the observed output when combining a file without the ABI attribute with one that has it set to 'unknown', and two minor nits. If we resolve that question, it all looks good to me - thanks for your patience!

File: lld/test/ELF/riscv-attributes.s
Comment (ilovepi) [reply lld/test/ELF/riscv-attributes.s:335]: So, we probably need to double check w/ @MaskRay, but I believe we don't emit attributes w/ 0 value in general. I think the following snippet seems to indicate that from a few lines below the call to `mergeAtomics`. I assume that is because the default value of 0 is pretty much 'not used', so it isn't emitted normally, and since UNKOWN is 0, it just falls into that case.

```
size_t size = 5 + merged.vendor.size() + 1 + 5;
  for (auto &attr : merged.intAttr)
    if (attr.second != 0)
      size += getULEB128Size(attr.first) + getULEB128Size(attr.second);
```

I'll update the patch to address the nits, but I'll try to wait for Fangrui to confirm if thats correct about 0 valued attributes before proceeding.

Comment (asb) [reply lld/test/ELF/riscv-attributes.s]: The main reason it _might_ be helpful to differentiate between the attribute being present but UNKNOWN and the attribute not being present at all, is that if the attribute is there (and emitted when merging UNKNOWN with NONE) you at least know it's been processed by tools that were aware of the attribute. Is that useful? Not sure. It's probably not hugely important either way, but thought I'd check if this was by design or not.

**Comment by ilovepi** [comment]
@MaskRay @asb Are we happy w/ this patch? I think I've addressed most of the concerns raised thus far, but I want try and finish the attributes patches this week if possible, since I'm back from EuroLLVM + vacation.

**Review by asb — APPROVED** [review]
LGTM. Thanks!

**Comment by lukel97** [comment]
Hi, I'm seeing a segfault with GNU ld after this when running clang with `-march=rv64g` and riscv64-linux-gnu-ld 2.42. With LLD 18.1.3 I get a warning:
```
ld.lld: warning: /var/folders/yl/qbdtspsx7t97vqcl6dm6q9540000gn/T/--4b08f5.o:(.riscv.attributes): invalid tag 0xe at offset 0x45
```

I haven't tried it with a ToT lld yet but I thought I'd flag this in case something was accidentally made backwards incompatible!

**Comment by asb** [comment]
Thanks Luke. I can reproduce locally (with GNU ld 2.42) so have gone ahead and reverted so we can decide what to do without breaking people's workflows in the meantime.

CC @kito-cheng 

It seems like there are two problems here, the first obviously by far the biggest:
* There seems to be a bug in GNU ld's elf attribute handling code that causes it to segfault when seeing an unrecognized attribute.
* Arguably warning whenever an unrecognized tag is encountered (as ld.lld does, via `ELFAttributeParser::parseAttributeList`) is overly verbose.

**Comment by ilovepi** [comment]
@lukel97 Thanks for the report and @asb thanks for the reverts. 

It does look like we’ll need to wait for bfd to catch up or at least not crash.

File: lld/ELF/Arch/RISCV.cpp
Comment (MaskRay) [inline lld/ELF/Arch/RISCV.cpp:1106]: This should use `errorOrWarn`.

Sorry that I did not follow up the discussion in time, but I'd appreciate that I had the opportunity to give this comment after the patch was approved by @asb ...

File: lld/test/ELF/riscv-attributes.s
Comment (MaskRay) [inline lld/test/ELF/riscv-attributes.s:52]: We don't use double blank lines

File: lld/test/ELF/riscv-attributes.s
Comment (MaskRay) [inline lld/test/ELF/riscv-attributes.s:54]: excess space

File: lld/test/ELF/riscv-attributes.s
Comment (MaskRay) [inline lld/test/ELF/riscv-attributes.s:47]: Mismatched

Upper case. And there is a better word than "merging"

File: lld/ELF/Arch/RISCV.cpp
Comment (ilovepi) [reply lld/ELF/Arch/RISCV.cpp:1106]: Noted. I'll follow up in https://github.com/llvm/llvm-project/pull/90266. I based this on your comment to a similar line in https://github.com/llvm/llvm-project/pull/84598#discussion_r1521836857, but it seems like I misunderstood?

**Comment by MaskRay** [comment]
> @MaskRay @asb Are we happy w/ this patch? I think I've addressed most of the concerns raised thus far, but I want try and finish the attributes patches this week if possible, since I'm back from EuroLLVM + vacation.

I am sorry it took a long time for me to notice the discussion and make some comments about the linker code.
While I appreciate the LGTM, given the number of linker comments I had to give,  landing this 7 hours after LGTM feels a bit rushed.

**Comment by ilovepi** [comment]
> > @MaskRay @asb Are we happy w/ this patch? I think I've addressed most of the concerns raised thus far, but I want try and finish the attributes patches this week if possible, since I'm back from EuroLLVM + vacation.
> 
> I am sorry it took a long time for me to notice the discussion and make some comments about the linker code. While I appreciate the LGTM, given the number of linker comments I had to give, landing this 7 hours after LGTM feels a bit rushed.

Apologies. That wasn't my intent. I'll be sure to allow some more time in the future.

**Comment by asb** [comment]
> > @MaskRay @asb Are we happy w/ this patch? I think I've addressed most of the concerns raised thus far, but I want try and finish the attributes patches this week if possible, since I'm back from EuroLLVM + vacation.
> 
> I am sorry it took a long time for me to notice the discussion and make some comments about the linker code. While I appreciate the LGTM, given the number of linker comments I had to give, landing this 7 hours after LGTM feels a bit rushed. (If asb's LGTM was given on the day you commented, I'd certainly not oppose as it was my fault not noticing the change in 3 days.)

Sorry for that @MaskRay. I should have given an LGTM conditional on you or another lld dev explicitly lgtming the linker code. I think given jrtc27 had been involved in the thread and a discussion thread with you was mentioned I'd misjudged the level of attention this had got already from lld folks - so my bad.

**Pull request merged**
